File: src/validateBRPhoneNumber.ts
const defaultErrorMsg = ['Invalid value passed', 'Invalid phone number', 'Unknown error'];
/**
 * @param {string} phoneNumber
 * @param {string[]} [errorMsg=defaultErrorMsg] optional
 * @example validateBRPhoneNumber('(11) 98765-4321');
 * @example validateBRPhoneNumber('(11) 98765-4321', ['Invalid phone number', 'Invalid format', 'Unknown error']);
 * @description This function returns three errors in the following order:
 *
 * Default:
 * ['Invalid value passed', 'Invalid phone number', 'Unknown error']
 *
 * Create a list of errors separated by commas in strings
 * @returns {object} An object with 'isValid' (boolean) and 'errorMsg' (string) properties.
 */
function validateBRPhoneNumber(phoneNumber: string, errorMsg = defaultErrorMsg) {
  if (typeof phoneNumber !== 'string') throw new TypeError('The input should be a string.');
  // Check to see if the passed error messages are valid; otherwise, return an error
  if (errorMsg) {
    if (!Array.isArray(errorMsg)) throw new Error('errorMsg must be an Array');
    for (let index = 0; index < errorMsg.length; index += 1) {
      if (errorMsg[index] != null && typeof errorMsg[index] !== 'string') {
        throw new TypeError('All values within the array must be strings or null/undefined.');
      }
    }
  }

  // Internal function to get the error message
  function getErrorMessage(index: number) {
    if (errorMsg && index >= 0 && index < errorMsg.length && errorMsg[index] != null) {
      return errorMsg[index];
    }
    return defaultErrorMsg[index];
  }
  if (!phoneNumber) {
    return {
      isValid: false,
      errorMsg: getErrorMessage(0),
    };
  }
  // Regular expression to validate Brazilian phone numbers
  const brPhoneNumberRegex = /^\(\d{2}\) \d{5}-\d{4}$/;
  try {
    if (!brPhoneNumberRegex.test(phoneNumber)) {
      return {
        isValid: false,
        errorMsg: getErrorMessage(1),
      };
    }
    return {
      isValid: true,
      errorMsg: null,
    };
  } catch (error) {
    return {
      isValid: false,
      errorMsg: getErrorMessage(2),
    };
  }
}
export default validateBRPhoneNumber;

File: packages/python/src/multiform_validator/validateBRPhoneNumber.py
import re

default_error_msg = ['Invalid value passed', 'Invalid phone number', 'Unknown error']

def validateBRPhoneNumber(phone_number, errorMsg=default_error_msg):
    if not isinstance(phone_number, str):
        raise TypeError('The input should be a string.')

    if errorMsg:
        if not isinstance(errorMsg, list):
            raise TypeError('errorMsg must be a list')
        for msg in errorMsg:
            if msg is not None and not isinstance(msg, str):
                raise TypeError('All values within the list must be strings or None.')

    def get_error_message(index):
        error_message = errorMsg[index]
        return error_message if error_message is not None else default_error_msg[index]

    if not phone_number:
        return {
            'isValid': False,
            'errorMsg': get_error_message(0),
        }

    br_phone_number_regex = r'^\(?\d{2}\)?[\s-]?9?\d{4}-?\d{4}$'
    if not re.match(br_phone_number_regex, phone_number):
        return {
            'isValid': False,
            'errorMsg': get_error_message(1),
        }

    return {
        'isValid': True,
        'errorMsg': None,
    }
